src/v/cloud_storage/remote_partition.cc

```diff
@@ -362,14 +362,22 @@ ss::future<> remote_partition::run_eviction_loop() {
     vlog(_ctxlog.debug, "remote partition eviction loop stopped");
 }
 
-void remote_partition::gc_stale_materialized_segments() {
+void remote_partition::gc_stale_materialized_segments(bool force_collection) {
+    // The remote_segment instances are materialized on demand. They are
+    // collected after some period of inactivity.
+    // To prevent high memory consumption in some corner cases the
+    // materialization of the new remote_segment triggers GC. The idea is
+    // that remote_partition should have only one remote_segment in materialized
+    // state when it's constantly in use and zero if not in use.
     vlog(
       _ctxlog.debug,
       "collecting stale materialized segments, {} segments materialized, {} "
       "segments total",
       _materialized.size(),
       _segments.size());
-    auto now = ss::lowres_clock::now();
+
+    auto now = ss::lowres_clock::now()
+               + (force_collection ? stm_max_idle_time : 0ms);
```

Comment:
nit: it's not a huge deal, but ti would be easier to reason about the logic here if `now` had the expected meaning and idle time was factored in through a separate variable as needed in comparison below.

Reply:
fixed